lib/r10k/util/purgeable.rb

```diff
@@ -38,23 +44,78 @@ def current_contents(recurse)
         end
       end
 
+      # @deprecated Unused helper function
+      #
       # @return [Array<String>] Directory contents that are expected but not present
       def pending_contents(recurse)
         desired_contents - current_contents(recurse)
       end
 
+      def matches?(test, path)
+        if test == path
+          true
+        elsif File.fnmatch?(test, path, FN_MATCH_OPTS)
+          true
+        else
+          false
+        end
+      end
+
+      # A method to collect potentially purgeable content without searching into
+      # ignored directories when recursively searching.
+      #
+      # @param dir [String, Pathname] The directory to search for purgeable content
+      # @param exclusion_gobs [Array<String>] A list of file paths or File globs
+      #     to exclude from recursion (These are generated by the classes that
+      #     mix this module into them and are typically programatically generated)
+      # @param allowed_gobs [Array<String>] A list of file paths or File globs to exclude
+      #     from recursion (These are passed in by the caller of purge! and typically
+      #     are user supplied configuration values)
+      # @param desireds_not_to_recurse_into [Array<String>] A list of file paths not to
+      #     recurse into. These are programatically generated, these exist to maintain
+      #     backwards compatibility with previous implementations that used File.globs
+      #     for "recursion", ie "**/{*,.[^.]*}" which would not recurse into dot directories.
+      # @param recurse [Boolean] Whether or not to recurse into child directories that do
+      #     not match other filters.
+      #
+      # @return [Array<String>] Contents which may be purged.
+      def potentially_purgeable(dir, exclusion_globs, allowed_globs, desireds_not_to_recurse_into, recurse)
+        children = Pathname.new(dir).children.reject do |path|
+          path = path.to_s
+
+          if exclusion_match = exclusion_globs.find { |exclusion| matches?(exclusion, path) }
+            logger.debug2 _("Not purging %{path} due to internal exclusion match: %{exclusion_match}") % {path: path, exclusion_match: exclusion_match}
+          elsif allowlist_match = allowed_globs.find { |allowed| matches?(allowed, path) }
+            logger.debug _("Not purging %{path} due to whitelist match: %{allowlist_match}") % {path: path, allowlist_match: allowlist_match}
+          elsif desired_match = desireds_not_to_recurse_into.grep(path).first
```

Comment:
Could this just be
```
else
  desired_match = desireds ...
end
```
I think that might be a little clearer

Reply:
I get the instinct to match the pattern of the other cases, which do assignment in the check, but that's sometimes frowned upon in itself, so doing it here where there is no body to the check seems even less good.

Reply:
Yeah, I thought I pushed that change up before lunch. Sorry.